rule htseq_count:
    "Count nb of reads in genes (at exon level)"
    input:
        sam = "data/align/{prefix}.Aligned.out.sam",
        gene_gff = config["gene_gff"]
    output:
        "data/count/{sample}.int"
    conda:
        "rules/count/htseq_env.yaml"
    shell:
        "htseq-count -s config[stranded] -i gene -t exon {input.sam} {input.gene_gff} > {output[0]}"
